<a href="https://colab.research.google.com/github/Diante830/Capstone-Project-1-Unemployment-Trends-in-Canada-A-Provincial-and-Demographic-Analysis/blob/main/Capstone_Project_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Unemployment Trends in Canada: A Provincial and Demographic Analysis

### Data Analytics & Visualization Project

This project analyzes unemployment trends in Canada using historical labour market data. The analysis focuses on understanding long-term changes in unemployment across provinces and demographic groups, and identifying patterns that reflect broader economic conditions.

Understanding unemployment trends helps provide insight into the health of the labour market, regional economic differences, and how economic shocks affect employment across the country.



---



## Introduction

The unemployment rate is one of the most important indicators used to evaluate the health of a country's labour market. In Canada, unemployment rates vary across provinces and demographic groups due to differences in economic structure, industry composition, and population characteristics.

Major economic events, such as recessions or the COVID-19 pandemic, can have significant impacts on employment levels across the country. Understanding these trends helps reveal long-term labour market patterns and regional disparities.

By analyzing historical unemployment data, we can better understand how employment conditions have evolved over time and which population groups may be most affected by economic changes.

## Project Objectives

The objective of this project is to analyze unemployment trends in Canada using historical labour market data from 1976 to the present.

This project aims to:

- Examine long-term unemployment trends at national and provincial levels
- Compare unemployment patterns across Canadian provinces
- Identify demographic differences in unemployment rates over time
- Use data visualizations to clearly communicate key trends and insights

## Dataset Overview

The dataset used in this analysis contains historical labour market statistics for Canada, including unemployment, employment, labour force participation, and population estimates across provinces and demographic groups.

The data was obtained from **:contentReference[oaicite:0]{index=0}**, which provides official labour market statistics for Canada.

### Dataset Characteristics

- **Data size:** 38,985 rows × 13 columns  
- **Time period:** 1976 to present  
- **Data types:** Combination of numerical (float64) and categorical (object) variables  
- **Geographic coverage:** Canadian provinces  

### Key Variables Used in the Analysis

For this project, several variables are used to analyze unemployment trends across provinces and demographic groups:

- **REF_DATE** – Date of the observation  
- **GEO** – Province or geographic region  
- **Age group** – Age demographic category  
- **Labour force** – Total number of individuals in the labour market  
- **Population** – Total population for the province and demographic group  
- **Unemployment rate** – Percentage of the labour force that is unemployed

In [ ]:
import pandas as pd

df = pd.read_csv('Unemployment_Canada_1976_present.csv')
df

,REF_DATE,GEO,Sex,Age group,Employment,Full-time employment,Labour force,Part-time employment,Population,Unemployment,Employment rate,Participation rate,Unemployment rate
0,1976-01,Alberta,Both sexes,15 to 24 years,231800.0,174900.0,252300.0,56900.0,362300.0,20500.0,64.0,69.6,8.1
1,1976-01,Alberta,Both sexes,15 to 64 years,802400.0,682100.0,837500.0,120300.0,1154800.0,35000.0,69.5,72.5,4.2
2,1976-01,Alberta,Both sexes,15 years and over,819500.0,693700.0,856500.0,125800.0,1276700.0,37000.0,64.2,67.1,4.3
3,1976-01,Alberta,Both sexes,25 to 54 years,491400.0,439800.0,505800.0,51600.0,661700.0,14400.0,74.3,76.4,2.8
4,1976-01,Alberta,Both sexes,25 years and over,587700.0,518800.0,604200.0,68900.0,914400.0,16500.0,64.3,66.1,2.7
...,...,...,...,...,...,...,...,...,...,...,...,...,...
38980,2023-01,Saskatchewan,Both sexes,15 to 64 years,553900.0,458100.0,579800.0,95800.0,719100.0,25900.0,77.0,80.6,4.5
38981,2023-01,Saskatchewan,Both sexes,15 years and over,589000.0,479000.0,615500.0,110000.0,912000.0,26500.0,64.6,67.5,4.3
38982,2023-01,Saskatchewan,Both sexes,25 to 54 years,379900.0,337300.0,394500.0,42600.0,442400.0,14600.0,85.9,89.2,3.7
38983,2023-01,Saskatchewan,Both sexes,25 years and over,507400.0,436000.0,526200.0,71500.0,774200.0,18700.0,65.5,68.0,3.6


In [ ]:
df.head()

,REF_DATE,GEO,Sex,Age group,Employment,Full-time employment,Labour force,Part-time employment,Population,Unemployment,Employment rate,Participation rate,Unemployment rate
0,1976-01,Alberta,Both sexes,15 to 24 years,231800.0,174900.0,252300.0,56900.0,362300.0,20500.0,64.0,69.6,8.1
1,1976-01,Alberta,Both sexes,15 to 64 years,802400.0,682100.0,837500.0,120300.0,1154800.0,35000.0,69.5,72.5,4.2
2,1976-01,Alberta,Both sexes,15 years and over,819500.0,693700.0,856500.0,125800.0,1276700.0,37000.0,64.2,67.1,4.3
3,1976-01,Alberta,Both sexes,25 to 54 years,491400.0,439800.0,505800.0,51600.0,661700.0,14400.0,74.3,76.4,2.8
4,1976-01,Alberta,Both sexes,25 years and over,587700.0,518800.0,604200.0,68900.0,914400.0,16500.0,64.3,66.1,2.7


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 38985 entries, 0 to 38984
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   REF_DATE               38985 non-null  object 
 1   GEO                    38985 non-null  object 
 2   Sex                    38985 non-null  object 
 3   Age group              38985 non-null  object 
 4   Employment             38985 non-null  float64
 5   Full-time employment   37290 non-null  float64
 6   Labour force           38985 non-null  float64
 7   Part-time employment   37289 non-null  float64
 8   Population             38985 non-null  float64
 9   Unemployment           38979 non-null  float64
 10  Employment rate        38985 non-null  float64
 11  Participation rate     38985 non-null  float64
 12  Unemployment rate      38979 non-null  float64
dtypes: float64(9), object(4)
memory usage: 3.9+ MB


In [ ]:
print(df.describe())

         Employment  Full-time employment  Labour force  \
count  3.898500e+04          3.729000e+04  3.898500e+04   
mean   1.674920e+06          1.402672e+06  1.817379e+06   
std    3.157376e+06          2.709031e+06  3.406108e+06   
min    5.000000e+03          4.100000e+03  5.700000e+03   
25%    1.428000e+05          1.039000e+05  1.605000e+05   
50%    4.172000e+05          3.123000e+05  4.475000e+05   
75%    1.643800e+06          1.316950e+06  1.787500e+06   
max    2.003230e+07          1.643700e+07  2.107830e+07   

       Part-time employment     Population  Unemployment  Employment rate  \
count           3.728900e+04  3.898500e+04  3.897900e+04     38985.000000   
mean            2.846479e+05  2.672215e+06  1.424808e+05        57.681218   
std             5.369204e+05  4.817956e+06  2.609124e+05        16.254158   
min             5.000000e+02  1.770000e+04  2.000000e+02        11.600000   
25%             2.420000e+04  2.993000e+05  1.480000e+04        50.900000   
50%   

# Data Cleaning

Before performing exploratory data analysis, the dataset is cleaned to ensure consistency, accuracy, and readability.

The cleaning process includes:

1. Renaming columns for clarity  
2. Dropping irrelevant columns  
3. Fixing column text and data types  
4. Handling missing values  
5. Preparing columns for analysis and visualization

### Step 1: Rename Columns

- Column names are simplified for readability and easier referencing in code.
- Spaces and special characters are removed.

In [ ]:
df = df.rename(columns={
    "REF_DATE": "year",
    "GEO": "province",
    "Sex": "sex",
    "Age group": "age_group",
    "Employment": "employment",
    "Full-time employment": "full_time",
    "Part-time employment": "part_time",
    "Labour force": "labour_force",
    "Population": "population",
    "Unemployment": "unemployed",
    "Employment rate": "employment_rate",
    "Participation rate": "participation_rate",
    "Unemployment rate": "unemployment_rate"
})

df.columns

Index(['year', 'province', 'sex', 'age_group', 'employment', 'full_time',
       'labour_force', 'Part-time employment ', 'population', 'unemployed',
       'employment_rate', 'participation_rate', 'unemployment_rate'],
      dtype='object')

### Step 2: Clean Text Columns

- Standardize text formatting using `strip()` and `title()`  
- Replace "Both Sexes" with "All" in `sex` column  
- Standardize age group labels for readability in plots  
- Remove extra spaces in `province` names

In [ ]:
# Cleaning the columns
df['sex'] = df['sex'].str.strip().str.title()
df['sex'] = df['sex'].replace({'Both Sexes': 'All'})

df['province'] = df['province'].str.strip().str.title()

df['age_group'] = df['age_group'].str.strip().str.title()
df['age_group'] = df['age_group'].replace({
    '15 To 24 Years': '15-24',
    '25 To 54 Years': '25-54',
    '55 To 64 Years': '55-64',
    '15 To 64 Years': '15-64',
    '15 Years And Over': '15+',
    '25 Years And Over': '25+',
    '55 Years And Over': '55+',
    '15 To 19 Years': '15-19',
    '20 To 24 Years': '20-24'
})

print("Sex categories:", df['sex'].unique())
print("Province categories:", df['province'].unique())
print("Age groups:", df['age_group'].unique())

Sex categories: ['All']
Province categories: ['Alberta' 'British Columbia' 'Canada' 'Manitoba' 'New Brunswick'
 'Newfoundland And Labrador' 'Nova Scotia' 'Ontario'
 'Prince Edward Island' 'Quebec' 'Saskatchewan']
Age groups: ['15-24' '15-64' '15+' '25-54' '25+' '55+' '15-19' '20-24' '55-64']


### Step 3: Fix Data Types & Convert Rate Columns

- Convert `year` to numeric for plotting trends over time.  
- Convert rate columns (`unemployment_rate`, `employment_rate`) from decimals to percentages for clarity in analysis and visualization.  
- Remove rows with missing `unemployment_rate` to maintain accuracy.

In [ ]:
df['year'] = pd.to_datetime(df['year']).dt.year
print(df['year'].head())

0    1976
1    1976
2    1976
3    1976
4    1976
Name: year, dtype: int32


In [ ]:
rate_cols = [
    'unemployment_rate',
    'employment_rate'
]
df[rate_cols] = df[rate_cols]
print(df[rate_cols].head())

   unemployment_rate  employment_rate
0                8.1             64.0
1                4.2             69.5
2                4.3             64.2
3                2.8             74.3
4                2.7             64.3


In [ ]:
df = df.dropna(subset=['unemployment_rate'])

print(df['unemployment_rate'].isna().sum())

0


### Step 4: Drop Unused Columns & Final Dataset Check

- Keep only the columns needed for this analysis:  
  `year`, `province`, `sex`, `age_group`, `unemployed`, `labour_force`, `population`, `unemployment_rate`  
- Drop columns that are not relevant (`employment`, `full_time`, `part_time`, `participation_rate`)  
- Verify the dataset is clean and ready for analysis using:
  - `df.head()` for a quick preview  
  - `df.info()` to check rows, columns, and data types  
  - `df.describe()` for summary statistics

In [ ]:
df = df[['year', 'province', 'sex', 'age_group', 'unemployed',
         'labour_force', 'population', 'unemployment_rate']]
print(df.head())

   year province  sex age_group  unemployed  labour_force  population  \
0  1976  Alberta  All     15-24     20500.0      252300.0    362300.0   
1  1976  Alberta  All     15-64     35000.0      837500.0   1154800.0   
2  1976  Alberta  All       15+     37000.0      856500.0   1276700.0   
3  1976  Alberta  All     25-54     14400.0      505800.0    661700.0   
4  1976  Alberta  All       25+     16500.0      604200.0    914400.0   

   unemployment_rate  
0                8.1  
1                4.2  
2                4.3  
3                2.8  
4                2.7  


In [ ]:
#Checking to see if missing enteries were dropped (confirmed 38984 from 38979)
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 38979 entries, 0 to 38984
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   year                   38979 non-null  int32  
 1   province               38979 non-null  object 
 2   sex                    38979 non-null  object 
 3   age_group              38979 non-null  object 
 4   employment             38979 non-null  float64
 5   full_time              37284 non-null  float64
 6   labour_force           38979 non-null  float64
 7   Part-time employment   37283 non-null  float64
 8   population             38979 non-null  float64
 9   unemployed             38979 non-null  float64
 10  employment_rate        38979 non-null  float64
 11  participation_rate     38979 non-null  float64
 12  unemployment_rate      38979 non-null  float64
dtypes: float64(9), int32(1), object(3)
memory usage: 4.0+ MB


In [ ]:
df.head()

,year,province,sex,age_group,employment,full_time,labour_force,Part-time employment,population,unemployed,employment_rate,participation_rate,unemployment_rate
0,1976,Alberta,All,15-24,231800.0,174900.0,252300.0,56900.0,362300.0,20500.0,6400.0,69.6,810.0
1,1976,Alberta,All,15-64,802400.0,682100.0,837500.0,120300.0,1154800.0,35000.0,6950.0,72.5,420.0
2,1976,Alberta,All,15+,819500.0,693700.0,856500.0,125800.0,1276700.0,37000.0,6420.0,67.1,430.0
3,1976,Alberta,All,25-54,491400.0,439800.0,505800.0,51600.0,661700.0,14400.0,7430.0,76.4,280.0
4,1976,Alberta,All,25+,587700.0,518800.0,604200.0,68900.0,914400.0,16500.0,6430.0,66.1,270.0


In [ ]:
df.describe()

,year,employment,full_time,labour_force,Part-time employment,population,unemployed,employment_rate,participation_rate,unemployment_rate
count,38985.000000,3.898500e+04,3.729000e+04,3.898500e+04,3.728900e+04,3.898500e+04,3.897900e+04,38985.000000,38985.000000,38979.000000
mean,1999.042478,1.674920e+06,1.402672e+06,1.817379e+06,2.846479e+05,2.672215e+06,1.424808e+05,5768.121842,63.677543,947.213115
std,13.590318,3.157376e+06,2.709031e+06,3.406108e+06,5.369204e+05,4.817956e+06,2.609124e+05,1625.415784,17.256123,456.519154
min,1976.000000,5.000000e+03,4.100000e+03,5.700000e+03,5.000000e+02,1.770000e+04,2.000000e+02,1160.000000,13.900000,80.000000
25%,1987.000000,1.428000e+05,1.039000e+05,1.605000e+05,2.420000e+04,2.993000e+05,1.480000e+04,5090.000000,59.300000,610.000000
50%,1999.000000,4.172000e+05,3.123000e+05,4.475000e+05,7.150000e+04,6.890000e+05,3.690000e+04,6040.000000,66.500000,850.000000
75%,2011.000000,1.643800e+06,1.316950e+06,1.787500e+06,2.924000e+05,2.698000e+06,1.502000e+05,6890.000000,75.500000,1200.000000
max,2023.000000,2.003230e+07,1.643700e+07,2.107830e+07,3.677100e+06,3.207040e+07,2.691100e+06,8780.000000,91.600000,3620.000000


# Exploratory Data Analysis (EDA) — Summary Tables

This analysis examines key statistics for Canadian unemployment to establish a foundational understanding of the dataset.  

We focus on:  
- **Overall unemployment rate summary** – Understand the general level and variation of unemployment across Canada.  
- **Records per province** – Examine how data is distributed across provinces and regions.  
- **Average unemployment rate by province** – Identify provinces with higher or lower unemployment on average.  
- **Average unemployment rate by year** – Observe how unemployment trends have changed over time.  
- **Average unemployment rate by age group** – Highlight demographic differences in unemployment.

These tables provide a baseline view of the data before deeper analysis and visualizations.

In [ ]:
print("Overall Unemployment Rate Summary:")
print(df['unemployment_rate'].describe())

print("\nNumber of records per province:")
print(df['province'].value_counts())

province_rates = (
    df.groupby('province')['unemployment_rate']
      .mean()
      .reset_index()
)

province_rates['unemployment_rate'] = province_rates['unemployment_rate'].round(1).astype(str) + '%'

print("\nAverage Unemployment Rate by Province:")
print(province_rates)


yearly_trends = (
    df.groupby('year')['unemployment_rate']
      .mean()
      .reset_index()
)

print("\nAverage Unemployment Rate by Year:")
print(yearly_trends.head())


age_group_rates = (
    df.groupby('age_group')['unemployment_rate']
      .mean()
      .reset_index()
)

print("\nAverage Unemployment Rate by Age Group:")
print(age_group_rates)

Overall Unemployment Rate Summary:
count    38979.000000
mean         9.472131
std          4.565192
min          0.800000
25%          6.100000
50%          8.500000
75%         12.000000
max         36.200000
Name: unemployment_rate, dtype: float64

Number of records per province:
province
Canada                       5085
British Columbia             3390
Manitoba                     3390
Ontario                      3390
New Brunswick                3390
Newfoundland And Labrador    3390
Nova Scotia                  3390
Quebec                       3390
Alberta                      3388
Prince Edward Island         3388
Saskatchewan                 3388
Name: count, dtype: int64

Average Unemployment Rate by Province:
                     province unemployment_rate
0                     Alberta              6.7%
1            British Columbia              8.4%
2                      Canada              9.5%
3                    Manitoba              6.5%
4               New Brunswi

#### Overall Unemployment Analysis

Summary of key trends in Canadian unemployment, looking at overall statistics, provincial patterns, yearly trends, and age group differences.  

Key insights include:  
1. Median unemployment rate is 8.5%, indicating periods of very high unemployment.  
2. Average unemployment rate is 9.47%, showing moderate long-term unemployment.  
3. Rates range from 0.8% to 36.2%, reflecting major economic shocks.  
4. Provinces such as Alberta, Saskatchewan, and Manitoba tend to have lower unemployment, while Atlantic provinces experience higher rates.  
5. Younger age groups (15–24) consistently face higher unemployment than adults (25–54).



---



# Exploratory Data Analysis (EDA) — Trends and Patterns

This analysis examines trends in unemployment to understand how they vary across provinces, over time, and among demographic groups.

Key areas of focus include:  
- **Provincial unemployment patterns** – How unemployment differs between provinces.  
- **Unemployment rates over time** – Observing historical trends and fluctuations.  
- **Population and labour force relationships** – How population size and labour force numbers relate to unemployment rates.  
- **Impact of major events** – For example, analyzing changes during the COVID‑19 pandemic.  

Descriptive statistics, aggregation, and grouping are used to uncover patterns that will inform subsequent visualizations and insights.

In [ ]:
province_rates = df.groupby('province')['unemployment_rate'].mean().reset_index()
province_rates['unemployment_rate'] = province_rates['unemployment_rate'].apply(lambda x: f"{x:.1f}%")
print(df['province'].value_counts())
print(province_rates)

province
Canada                       5085
British Columbia             3390
Manitoba                     3390
Ontario                      3390
New Brunswick                3390
Newfoundland And Labrador    3390
Nova Scotia                  3390
Quebec                       3390
Alberta                      3388
Prince Edward Island         3388
Saskatchewan                 3388
Name: count, dtype: int64
                     province unemployment_rate
0                     Alberta              6.7%
1            British Columbia              8.4%
2                      Canada              9.5%
3                    Manitoba              6.5%
4               New Brunswick             11.2%
5   Newfoundland And Labrador             16.0%
6                 Nova Scotia             10.3%
7                     Ontario              7.7%
8        Prince Edward Island             12.1%
9                      Quebec              9.7%
10               Saskatchewan              6.0%


In [ ]:
print(df['unemployment_rate'].describe())

count    38979.000000
mean         9.472131
std          4.565192
min          0.800000
25%          6.100000
50%          8.500000
75%         12.000000
max         36.200000
Name: unemployment_rate, dtype: float64


In [ ]:
province_year_rates = (
    df.groupby(['year', 'province'])['unemployment_rate']
      .mean()
      .reset_index()
)

print("\nUnemployment Rate Trends by Province and Year:")
print(
    province_year_rates.assign(
        unemployment_rate=province_year_rates['unemployment_rate']
        .round(1)
        .astype(str) + '%'
    )
)


Unemployment Rate Trends by Province and Year:
     year              province unemployment_rate
0    1976               Alberta              3.7%
1    1976      British Columbia              8.2%
2    1976                Canada              7.9%
3    1976              Manitoba              4.4%
4    1976         New Brunswick             10.4%
..    ...                   ...               ...
523  2023           Nova Scotia              5.8%
524  2023               Ontario              5.6%
525  2023  Prince Edward Island              8.8%
526  2023                Quebec              4.2%
527  2023          Saskatchewan              4.6%

[528 rows x 3 columns]


#### Unemployment Rate vs. Population and Labour Force

This section examines how population and labour force size relate to unemployment rates across Canadian provinces over time.  

Key points to consider:  
- **Population trends over time** can reveal demographic growth and potential pressure on the labour market.  
- **Large provinces** like Ontario have sizable populations and labour forces, but this does not always correspond to higher unemployment rates.  
- **Smaller provinces** may have smaller labour forces yet experience higher unemployment rates in certain years.  
- **Scatter plots** of labour force or population vs. unemployment rate can help visualize correlations and identify patterns.

In [ ]:
province_pop_lf_rates = (
    df.groupby(['year', 'province'])[['population', 'labour_force', 'unemployment_rate']]
      .mean()
      .reset_index()
)

province_pop_lf_rates['unemployment_rate'] = province_pop_lf_rates['unemployment_rate'].round(1).astype(str) + '%'
province_pop_lf_rates['population'] = province_pop_lf_rates['population'].apply(lambda x: f"{int(x):,}")
province_pop_lf_rates['labour_force'] = province_pop_lf_rates['labour_force'].apply(lambda x: f"{int(x):,}")

print(province_pop_lf_rates)

     year              province population labour_force unemployment_rate
0    1976               Alberta    803,773      552,029              3.7%
1    1976      British Columbia  1,134,011      717,550              8.2%
2    1976                Canada  7,586,305    4,758,568              7.9%
3    1976              Manitoba    443,395      280,952              4.4%
4    1976         New Brunswick    288,406      159,629             10.4%
..    ...                   ...        ...          ...               ...
523  2023           Nova Scotia    512,616      334,450              5.8%
524  2023               Ontario  7,652,933    5,251,950              5.6%
525  2023  Prince Edward Island     85,000       57,416              8.8%
526  2023                Quebec  4,355,050    2,987,333              4.2%
527  2023          Saskatchewan    552,883      389,500              4.6%

[528 rows x 5 columns]


In [ ]:
ontario_rates = province_year_rates[
    province_year_rates['province'] == 'Ontario'
]

print("\nOntario Unemployment Rate Over the Years:")
print(
    ontario_rates.assign(
        unemployment_rate=ontario_rates['unemployment_rate']
        .round(1)
        .astype(str) + '%'
    )
)


Ontario Unemployment Rate Over the Years:
     year province unemployment_rate
7    1976  Ontario              5.9%
18   1977  Ontario              6.8%
29   1978  Ontario              7.1%
40   1979  Ontario              6.4%
51   1980  Ontario              6.7%
62   1981  Ontario              6.4%
73   1982  Ontario              9.7%
84   1983  Ontario             10.3%
95   1984  Ontario              8.9%
106  1985  Ontario              7.9%
117  1986  Ontario              7.0%
128  1987  Ontario              6.1%
139  1988  Ontario              5.1%
150  1989  Ontario              5.0%
161  1990  Ontario              6.1%
172  1991  Ontario              9.6%
183  1992  Ontario             11.0%
194  1993  Ontario             11.1%
205  1994  Ontario              9.9%
216  1995  Ontario              9.1%
227  1996  Ontario              9.2%
238  1997  Ontario              8.8%
249  1998  Ontario              7.6%
260  1999  Ontario              6.7%
271  2000  Ontario              

#### Filter for COVID-19 years (2020–2023) for all provinces

In [ ]:
covid_impact_all = province_year_rates[
    province_year_rates['year'] >= 2020
]

print(
    covid_impact_all.assign(
        unemployment_rate=covid_impact_all['unemployment_rate']
        .round(1)
        .astype(str) + '%'
    )
)

     year                   province unemployment_rate  unemployment_rate_frac
484  2020                    Alberta             12.7%                0.127250
485  2020           British Columbia             10.1%                0.100639
486  2020                     Canada             12.7%                0.126769
487  2020                   Manitoba              8.8%                0.088292
488  2020              New Brunswick             11.4%                0.113528
489  2020  Newfoundland And Labrador             15.6%                0.156181
490  2020                Nova Scotia             10.6%                0.105944
491  2020                    Ontario             10.8%                0.108153
492  2020       Prince Edward Island             11.1%                0.111500
493  2020                     Quebec              9.8%                0.097778
494  2020               Saskatchewan              9.1%                0.090917
495  2021                    Alberta              9.

#### Age Group Unemployment Trends by Province

This analysis examines unemployment trends across different age groups in each province over time, highlighting both long-term patterns and the impact of major events such as the COVID‑19 pandemic.

Key observations:
- **Young workers (15–24)** consistently experience higher unemployment rates compared to older age groups.
- **Prime working-age and older groups (25–54, 55+)** generally show lower and more stable unemployment.
- Unemployment patterns vary by province, revealing important regional and demographic differences in labour market outcomes.

In [ ]:
age_province_rates = (
    df.groupby(['year', 'province', 'age_group'])['unemployment_rate']
      .mean()
      .reset_index()
)

age_province_rates['unemployment_rate'] = age_province_rates['unemployment_rate'].round(1).astype(str) + '%'
print(age_province_rates)

      year      province age_group unemployment_rate
0     1976       Alberta       15+              3.9%
1     1976       Alberta     15-24              7.3%
2     1976       Alberta     15-64              3.9%
3     1976       Alberta       25+              2.4%
4     1976       Alberta     25-54              2.5%
...    ...           ...       ...               ...
3307  2023  Saskatchewan     15-24              8.7%
3308  2023  Saskatchewan     15-64              4.5%
3309  2023  Saskatchewan       25+              3.6%
3310  2023  Saskatchewan     25-54              3.7%
3311  2023  Saskatchewan       55+              3.1%

[3312 rows x 4 columns]


In [ ]:
#for visuals
province_pop_lf_rates = (
    df.groupby(['year', 'province'])[['population', 'labour_force', 'unemployment_rate']]
      .mean()
      .reset_index()
)
print(province_pop_lf_rates)

     year              province    population  labour_force  unemployment_rate
0    1976               Alberta  8.037732e+05  5.520296e+05           3.718310
1    1976      British Columbia  1.134011e+06  7.175500e+05           8.222222
2    1976                Canada  7.586306e+06  4.758569e+06           7.864815
3    1976              Manitoba  4.433958e+05  2.809528e+05           4.443056
4    1976         New Brunswick  2.884069e+05  1.596292e+05          10.444444
..    ...                   ...           ...           ...                ...
523  2023           Nova Scotia  5.126167e+05  3.344500e+05           5.766667
524  2023               Ontario  7.652933e+06  5.251950e+06           5.650000
525  2023  Prince Edward Island  8.500000e+04  5.741667e+04           8.816667
526  2023                Quebec  4.355050e+06  2.987333e+06           4.250000
527  2023          Saskatchewan  5.528833e+05  3.895000e+05           4.650000

[528 rows x 5 columns]


#### Overall Unemployment Analysis

This analysis examines how population size and labour force relate to unemployment rates across Canadian provinces over time, highlighting regional differences, economic cycles, and pandemic impacts.  

Key insights include:  
1. Average unemployment rates differ across provinces, indicating persistent regional labour market differences.  
2. Year-to-year fluctuations reflect recessions and recoveries, which affect provinces differently.  
3. High unemployment is not always linked to population size — smaller provinces can experience higher unemployment, while larger provinces may have lower rates.  
4. Understanding scale: a 5% unemployment rate in a large province like Ontario affects far more people than the same rate in a smaller province.  
5. Most provinces recovered to near pre-pandemic levels by 2022–2023, though some Atlantic provinces experienced slower recovery.



---



# Visualizations

This section presents visual representations of Canadian unemployment trends to complement the analysis. The charts highlight patterns across provinces, age groups, and time, making it easier to interpret key insights and regional differences. Visualizations also illustrate the impact of major events, such as economic recessions and the COVID‑19 pandemic.

###Province-Level Averages for Unemployment Rates
*This interactive bar chart allows exploration of average unemployment rates across Canadian provinces.*

In [ ]:
from bokeh.plotting import figure, output_notebook, show
from bokeh.models import ColumnDataSource, NumeralTickFormatter
output_notebook()

# Prep
province_rates_plot = df.groupby('province')['unemployment_rate'].mean().reset_index()
province_rates_plot['unemployment_rate'] = province_rates_plot['unemployment_rate'] / 100

source = ColumnDataSource(province_rates_plot)

# Figure
p = figure(x_range=province_rates_plot['province'],
           height=400, width=700,
           title="Average Unemployment Rate by Province",
           x_axis_label='Province', y_axis_label='Unemployment Rate (%)',
           toolbar_location=None, tools="")

# V-Bars
p.vbar(x='province', top='unemployment_rate', width=0.8, source=source, color="skyblue")

# Customizations
p.xgrid.grid_line_color = None
p.y_range.start = 0
p.xaxis.major_label_orientation = 1.0

p.yaxis.formatter = NumeralTickFormatter(format="0.0%")

show(p)

#### Insight #1

Regional economic disparities, possibly due to differences in industry composition, resource dependence, or population size.

#### Insight #2

Provinces with high unemployment, like Newfoundland and Labrador, pull the national average up, even if other provinces have moderate rates.

#### Insight #3

Provinces with lower unemployment rates, like Alberta (6.7%) and Ontario (7.7%), likely have diverse economies or strong job markets.

#### Insight #4

Regional economic disparities, possibly due to differences in industry composition, resource dependence, or population size.






---



# Unemployment Rate Trends Per Year
*This interactive line chart showing trends over years. Hover shows exact unemployment rate per year.*

In [ ]:
from bokeh.models import HoverTool, ColumnDataSource, FixedTicker, NumeralTickFormatter
from bokeh.plotting import figure, show, output_notebook

output_notebook()

# Prep
yearly_trends['unemployment_rate_frac'] = yearly_trends['unemployment_rate'] / 100
yearly_trends_source = ColumnDataSource(yearly_trends)

# Figure
p = figure(height=400, width=800,
           title="Average Unemployment Rate in Canada Over Time",
           x_axis_label='Year', y_axis_label='Unemployment Rate',
           tools="pan,wheel_zoom,box_zoom,reset,save")

# Added lines and points
p.line(x='year', y='unemployment_rate_frac', source=yearly_trends_source, line_width=2, color='skyblue')
p.circle(x='year', y='unemployment_rate_frac', source=yearly_trends_source, size=6, color='blue')


hover = HoverTool(tooltips=[("Year", "@year"), ("Rate", "@unemployment_rate{0.1f}%")])
p.add_tools(hover)

p.yaxis.formatter = NumeralTickFormatter(format="0%")
p.yaxis.ticker = FixedTicker(ticks=[i/100 for i in range(0, 35, 2)])

custom_years = list(range(1975, 2026, 5))
p.xaxis.ticker = FixedTicker(ticks=custom_years)
p.xaxis.major_label_orientation = 1.0

show(p)

####Insight #1

1980s: Sharp rise to 12.3% in 1983-1984, likely reflecting the global recession and high interest rates.

####Insight #2

1990s: Peak of 12.7% in 1993, corresponding to the 90s recession. Then spike to 9.5% after the 2008 financial crisis.

####Insight #3

2020: Jump to 11.2% due to the COVID՞19 pandemic.

####Insight #4

These peaks are followed by gradual recoveries, seems to be a cyclical trend



---



#Comparing Unemployment Among Age Groups.
*Bar chart comparing unemployment among youth, adults, and seniors.*

In [ ]:
from bokeh.models import ColumnDataSource, FixedTicker, NumeralTickFormatter
from bokeh.plotting import figure, show, output_notebook

output_notebook()

# Prep
age_group_rates['unemployment_rate_frac'] = age_group_rates['unemployment_rate'] / 100
age_source = ColumnDataSource(age_group_rates)

# Figure
p = figure(x_range=age_group_rates['age_group'],
           height=400, width=700,
           title="Average Unemployment Rate by Age Group",
           x_axis_label='Age Group', y_axis_label='Unemployment Rate',
           toolbar_location=None, tools="")

# V-Bars
p.vbar(x='age_group', top='unemployment_rate_frac', width=0.8, source=age_source, color="purple")

# Remove x-grid
p.xgrid.grid_line_color = None
p.y_range.start = 0

# Percentage on y-axis
p.yaxis.formatter = NumeralTickFormatter(format="0%")
p.yaxis.ticker = FixedTicker(ticks=[i/100 for i in range(0, 35, 2)])

show(p)

#### Insight #1

Teenagers and young adults struggle more to find jobs, likely due to less work experience, part-time, or competition for entry-level roles.

#### Insight #2

Ages 25-54 have the lowest unemployment rates

• Ages 25+: 7.79%

• Ages 25-54: 7.83%

#### Insight #3

• Ages 55-64: 6.59% More stability for experienced workers or few people job seeking after retirement age.

#### Insight #4

Unemployment is highest among the youngest workers, and remains relatively low for older adults.



---



#Unemployment Trends For Each Province Over Time
*Interactive line chart showing unemployment rate trends by Canadian province over time.*

In [ ]:
from bokeh.plotting import figure, show, output_notebook
from bokeh.models import ColumnDataSource, HoverTool, FixedTicker, NumeralTickFormatter
from bokeh.palettes import Category10

output_notebook()

# Percentage format
province_year_rates['unemployment_rate_frac'] = (
    province_year_rates['unemployment_rate'] / 100
)


provinces = province_year_rates['province'].unique().tolist()
colors = Category10[10]

p = figure(
    height=700,
    width=1200,
    title="Unemployment Rate Trends by Province Over Time",
    x_axis_label="Year",
    y_axis_label="Unemployment Rate (%)",
    tools="pan,wheel_zoom,box_zoom,reset,save"
)

for i, province in enumerate(provinces):
    province_data = province_year_rates[
        province_year_rates['province'] == province
    ]

    source = ColumnDataSource(province_data)


    p.line(
        x='year',
        y='unemployment_rate_frac',
        source=source,
        color=colors[i % len(colors)],
        line_width=2,
        legend_label=province
    )


    p.circle(
        x='year',
        y='unemployment_rate_frac',
        source=source,
        color=colors[i % len(colors)],
        size=4
    )

hover = HoverTool(
    tooltips=[
        ("Province", "@province"),
        ("Year", "@year"),
        ("Unemployment Rate", "@unemployment_rate{0.1f}%")
    ]
)
p.add_tools(hover)

# 0%–36% in 2% increments
p.yaxis.ticker = FixedTicker(
    ticks=[i / 100 for i in range(0, 36, 2)]
)
p.yaxis.formatter = NumeralTickFormatter(format="0%")

# Every 5 years
p.xaxis.ticker = FixedTicker(ticks=list(range(1976, 2026, 5)))
p.xaxis.major_label_orientation = 1.0

p.legend.title = "Provinces"
p.legend.location = "top_right"
p.legend.click_policy = "hide"
p.legend.label_text_font_size = "8pt"
p.legend.glyph_height = 10
p.legend.glyph_width = 10
p.legend.spacing = 2
p.legend.padding = 5

show(p)

#Population and Labour Force For Each Province
*Interactive pie charts showing each province’s population share of Canada’s labour force through the years*

In [ ]:
from bokeh.plotting import figure, show, output_notebook
from bokeh.models import ColumnDataSource, HoverTool
from bokeh.palettes import Category10
from bokeh.transform import cumsum
import math

output_notebook()

pie_df = province_pop_lf_rates[
    province_pop_lf_rates['year'] == 2023
].copy()


pie_df['share'] = pie_df['labour_force'] / pie_df['labour_force'].sum()
pie_df['angle'] = pie_df['share'] * 2 * math.pi


pie_df['colors'] = [Category10[10][i % len(Category10[10])] for i in range(len(pie_df))]

source = ColumnDataSource(pie_df)


p = figure(
    height=500,
    width=500,
    title="Provincial Share of Canadian Labour Force (2023)",
    tools="hover",
    tooltips=[
        ("Province", "@province"),
        ("Labour Force", "@labour_force{0,0}"),
        ("Share", "@share{0.0%}")
    ]
)

p.wedge(
    x=0, y=0,
    radius=0.4,
    start_angle=cumsum('angle', include_zero=True),
    end_angle=cumsum('angle'),
    line_color="white",
    fill_color='colors',
    source=source
)


p.axis.visible = False
p.grid.grid_line_color = None

show(p)

In [ ]:
from bokeh.plotting import figure, show, output_notebook
from bokeh.models import ColumnDataSource, HoverTool
from bokeh.palettes import Category10
from bokeh.transform import cumsum
import math

output_notebook()

# Filter for one year
pie_df = province_pop_lf_rates[
    province_pop_lf_rates['year'] == 1976
].copy()

# Calculate share and angle using population instead of labour force
pie_df['share'] = pie_df['population'] / pie_df['population'].sum()
pie_df['angle'] = pie_df['share'] * 2 * math.pi

# Assign colors
pie_df['colors'] = [Category10[10][i % len(Category10[10])] for i in range(len(pie_df))]

source = ColumnDataSource(pie_df)

# Create figure
p = figure(
    height=500,
    width=500,
    title="Provincial Share of Canadian Population (1976)",
    tools="hover",
    tooltips=[
        ("Province", "@province"),
        ("Population", "@population{0,0}"),
        ("Share", "@share{0.0%}")
    ]
)

# Draw wedges
p.wedge(
    x=0, y=0,
    radius=0.4,
    start_angle=cumsum('angle', include_zero=True),
    end_angle=cumsum('angle'),
    line_color="white",
    fill_color='colors',
    source=source
)

# Hide axis and grid
p.axis.visible = False
p.grid.grid_line_color = None

show(p)


In [ ]:
from bokeh.plotting import figure, show, output_notebook
from bokeh.models import ColumnDataSource, HoverTool
from bokeh.palettes import Category10
from bokeh.transform import cumsum
import math

output_notebook()

pie_df = province_pop_lf_rates[
    province_pop_lf_rates['year'] == 2023
].copy()

# Calculate share and angle
pie_df['share'] = pie_df['population'] / pie_df['population'].sum()
pie_df['angle'] = pie_df['share'] * 2 * math.pi


pie_df['colors'] = [Category10[10][i % len(Category10[10])] for i in range(len(pie_df))]

source = ColumnDataSource(pie_df)


p = figure(
    height=500,
    width=500,
    title="Provincial Share of Canadian Labour Force (2023)",
    tools="hover",
    tooltips=[
        ("Province", "@province"),
        ("Labour Force", "@labour_force{0,0}"),
        ("Share", "@share{0.0%}")
    ]
)

p.wedge(
    x=0, y=0,
    radius=0.4,
    start_angle=cumsum('angle', include_zero=True),
    end_angle=cumsum('angle'),
    line_color="white",
    fill_color='colors',
    source=source
)


p.axis.visible = False
p.grid.grid_line_color = None

show(p)

#### Insight #1

Across all provinces, both population and labour force have increased significantly from 1976 to 2023.

#### Insight #2

• Example: Ontario grew from ֦1.13 million in 1976 to ֦7.65 million in 2023.

#### Insight #3

Ontario has 5.25M in the labour force out of 7.65M population Յ֦69%Ն, while Prince Edward Island has 57K out of 85K Յ֦67%Ն.

#### Insight #4

This shows that larger populations do not automatically guarantee proportional labour force growth.



---



#COVID-19 Unemployment Rate and Recovery Rate
*Interactive bar charts showing provincial unemployment rates in Canada during the COVID‑19 pandemic (2020–2023).*

In [ ]:
from bokeh.plotting import figure, show, output_notebook
from bokeh.models import ColumnDataSource, FactorRange, HoverTool
from bokeh.palettes import Category10, Category20

output_notebook()

# Prep
years = covid_impact_all['year'].unique().astype(str).tolist()
provinces = covid_impact_all['province'].unique().tolist()


factors = [(str(year), province) for year in years for province in provinces]


unemployment_values = covid_impact_all.sort_values(['year','province'])['unemployment_rate'].tolist()

if len(provinces) <= 10:
    colors = Category10[len(provinces)]
elif len(provinces) <= 20:
    colors = Category20[len(provinces)]
else:
    colors = [Category20[20][i % 20] for i in range(len(provinces))]

# Province Colours
color_list_for_source = [colors[provinces.index(prov)] for year, prov in factors]


source = ColumnDataSource(data=dict(
    x=factors,
    unemployment=unemployment_values,
    colors_for_bars=color_list_for_source
))

# Figure
p = figure(x_range=FactorRange(*factors), height=400, width=900,
           title="COVID-19 Era Unemployment by Province (2020–2023)",
           y_axis_label="Unemployment Rate (%)",
           toolbar_location=None, tools="")


p.vbar(x='x', top='unemployment', width=0.8, source=source, color='colors_for_bars')


p.xaxis.major_label_orientation = 1.2
p.yaxis.ticker = list(range(0, 36, 2))
p.yaxis.axis_label_standoff = 10


hover = HoverTool(tooltips=[("Year, Province", "@x"), ("Unemployment", "@unemployment{0.1f}%")])
p.add_tools(hover)


show(p)

#### Insight #1

The impact was uneven, with provinces dependent on tourism, hospitality, and resource sectors being most affected

#### Insight #2

By 2022՟2023, many provinces returned to prepandemic rates. This demonstrates a fast and strong post-pandemic labor market recovery

#### Insight #3

Regional inequalities are still shown, with Atlantic provinces lagging behind the national average even in 2023

#### Insight #4

Even in 2023, some provinces still struggle with higher unemployment:

• Newfoundland and Labrador: 12.5%

• Prince Edward Island: 8.8%

• New Brunswick: 8.4%



---



## Key Takeaways

### Age Gap in Job Security
Youth (15-24) face higher unemployment, likely due to limited experience and competition for entry-level positions.

### Workforce Planning Considerations
Analysis highlights the need to account for:
- Population growth and demographic shifts  
- Economic trends  
- Regional and age-group disparities  

### Impact of Economic Downturns
Major economic shocks are visible in the data, including:
- 1980s Recession  
- 2008 Financial Crisis  
- 2020 COVID-19 Pandemic  

### Uneven Pandemic Recovery
Recovery speed varied by province:
- Ontario, Quebec, and western provinces recovered fastest  
- Some Atlantic provinces continue to lag behind

#Thank You